src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java

```diff
@@ -0,0 +1,227 @@
+        @ConfigDefault("7")
+        int getMaximumRetries();
+
+        @Config("initial_retry_interval_millis")
+        @ConfigDefault("1000")
+        int getInitialRetryIntervalMillis();
+
+        @Config("maximum_retry_interval_millis")
+        @ConfigDefault("120000")
+        int getMaximumRetryIntervalMillis();
+
+        @Config("timeout_millis")
+        @ConfigDefault("60000")
+        int getTimeoutMills();
+    }
+
+    public enum Mode
+    {
+        INSERT,
+        REPLACE;
+
+        @JsonValue
+        @Override
+        public String toString()
+        {
+            return name().toLowerCase(Locale.ENGLISH);
+        }
+
+        @JsonCreator
+        public static Mode fromString(String value)
+        {
+            switch (value) {
+                case "insert":
+                    return INSERT;
+                case "replace":
+                    return REPLACE;
+                default:
+                    throw new ConfigException(String.format("Unknown mode '%s'. Supported modes are insert, truncate_insert, replace", value));
+            }
+        }
+    }
+
+    @Override  // Overridden from |OutputTaskValidatable|
+    public void validateOutputTask(PluginTask task, Schema embulkSchema, int taskCount)
+    {
+        if (task.getNodes().size() > 0) {
+            for (NodeAddressTask node : task.getNodes()) {
+                if (node.getHost().endsWith("es.amazonaws.com")) {
+                    throw new ConfigException("This plugin does't support AWS Elasticsearch Service.");
+                }
+            }
+        }
+
+        try (Jetty92RetryHelper retryHelper = createRetryHelper(task)) {
+            log.info(String.format("Connecting to Elasticsearch version:%s", client.getEsVersion(task, retryHelper)));
+
+            if (task.getMode().equals(Mode.REPLACE)) {
+                task.setAlias(Optional.of(task.getIndex()));
+                task.setIndex(client.generateNewIndexName(task.getIndex()));
+                if (client.isExistsIndex(task.getAlias().orNull(), task, retryHelper) && !client.isExistsAlias(task.getAlias().orNull(), task, retryHelper)) {
+                    throw new ConfigException(String.format("Invalid alias name [%s], an index exists with the same name as the alias", task.getAlias().orNull()));
+                }
+            }
+            log.info(String.format("Inserting data into index[%s]", task.getIndex()));
+        }
+    }
+
+    @Override  // Overridden from |ServiceRequestMapperBuildable|
+    public JacksonServiceRequestMapper buildServiceRequestMapper(PluginTask task)
+    {
+        return JacksonServiceRequestMapper.builder()
+                .add(new JacksonAllInObjectScope(), new JacksonTopLevelValueLocator("record"))
+                .build();
+    }
+
+    @Override  // Overridden from |RecordBufferBuildable|
+    public RecordBuffer buildRecordBuffer(PluginTask task)
+    {
+        return new JacksonTaskReportRecordBuffer("records");
```

Comment:
This `JacksonTaskReportRecordBuffer` just embeds all outputs from each Task in `TaskReport` on memory. Note that it can cause OOM if the output is so large. `JacksonTaskReportRecordBuffer` is just an example, and not recommended for large production usage.

It is expected to implement `RecordBuffer` per plugin for the time being. The base library does not contain better implementation for now since different destinations can have different buffering strategies. Once we have typical usages, we may import them into the library later.

* Some destinations may accept parallel uploads from each Task. Uploading from its own `RecordBuffer` may be fine for the destinations.
* Some destinations may accept very large ordered data in just one upload. Buffering into external storage, putting a pointer to the storage in `TaskReport`, and finally reading the external buffer from `transaction` may be required.

See the Javadoc below for the details:
https://github.com/embulk/embulk-base-restclient/blob/46c0ac6ddc5128ec2c4d4d52545b72f3f3c6b7df/src/main/java/org/embulk/base/restclient/record/RecordBuffer.java#L5-L21